### PR TITLE
feat(1951): New queue service implementation

### PR DESCRIPTION
**Pull request by pritamstyz4ever**
## Context

executor-queue is responsible for running periodic build jobs and enforcing freezeWindows and is now tightly coupled with Screwdriver API. It's important that API is kept light weight and heavy processing logic is split into separate micro services.

## Objective

This PR creates a new queue service repo and merges the executor-queue and queue-worker into microservices, thereby creating a decoupled system.

Design Doc:
https://github.com/screwdriver-cd/screwdriver/pull/1962

## References

https://github.com/screwdriver-cd/screwdriver/issues/1951

Other related PR's
https://github.com/screwdriver-cd/data-schema/pull/385
https://github.com/screwdriver-cd/executor-queue/pull/81
https://github.com/screwdriver-cd/screwdriver/pull/1962

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: bin/server [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+const server = require('../app');
+
+// Setup HTTPd
+const httpdConfig = config.get('httpd');
+// Special urls for things like the UI
+const ecosystem = config.get('ecosystem');
+
+const queueConfig = config.get('queue');
+
+// ecosystem.api = httpdConfig.uri;
```
Comment (tkyi) [inline bin/server]: commented-out code?

File: config/default.yaml [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  # https://github.com/screwdriver-cd/screwdriver/blob/095eaf03e053991443abcbde91c62cfe06a28cba/lib/server.js#L141
+  uri: http://IP_ADDRESS:PORT
+
+ecosystem:
+  # Externally routable URL for the User Interface
+  ui: http://IP_ADDRESS:PORT
+  # Externally routable URL for the API
+  api: http://IP_ADDRESS:PORT
+  # Externally routable URL for the Artifact Store
+  store: http://http://IP_ADDRESS:PORT
```
Comment (jithine) [inline config/default.yaml]: typo ? 

Also missing  `custom-environment-variables.yaml` like what we have in `queue-worker` https://github.com/screwdriver-cd/queue-worker/blob/master/config/custom-environment-variables.yaml

File: lib/queue.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+     * @param  {Object}         config.redisConnection      Connection details for redis
+     * @param  {String}         [config.prefix]             Prefix for queue name
+     * @param  {Object}         [config.breaker]            Optional breaker config
+     */
+    constructor(config = {}) {
+        if (!config.redisConnection) {
+            throw new Error('No redis connection passed in');
+        }
+
+        const breaker = Object.assign({}, config.breaker || {});
```
Comment (jithine) [inline lib/queue.js]: call it `breakerOptions`

File: lib/server.js [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        await registerPlugins(server, config);
+
+        await server.register({ plugin: Blipp, options: { showAuth: true } });
+
+        await server.start();
+        logger.info('Server running on %s', server.info.uri);
+
+        // init queue
+        const res1 = await server.inject({
+            url: '/v1/queue/message',
```
Comment (jithine) [inline lib/server.js]: don't we need a `/status` end point as well ?

File: plugins/helper.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+ * @param {String} buildId
+ * @param {String} token
+ * @param {String} status
+ * @param {String} statusMessage
+ * @param {String} apiUri
+ * @param {Object} updateConfig
+ */
+async function updateBuildStatusWithRetry(updateConfig) {
+    const { buildId, token, status, statusMessage, apiUri } = updateConfig;
+    // const buildConfig = await redisInstance.hget(`${queuePrefix}buildConfigs`, buildId)
```
Comment (jithine) [inline plugins/helper.js]: commented out code ? Also there there seems to be common options for requestrestry which could be defined in a global constant.

File: screwdriver.yaml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            screwdriver.cd/cpu: TURBO
+        requires:
+            - ~pr
+            - ~commit
+            - ~sd@73:publish #artifact-bookend https://cd.screwdriver.cd/pipelines/73
+        steps:
+            - setup-ci: git clone https://github.com/screwdriver-cd/toolbox.git ci
+            - install: npm install
+            - duplicate: NPM_FILTER=screwdriver- ./ci/npm-dups.sh
+            # - test: npm test
```
Comment (jithine) [inline screwdriver.yaml]: we should add these back and let npm handle it.

File: screwdriver.yaml [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            - K8S_TOKEN
+
+    # Deploy to our prod environment and run tests
+    prod:
+        requires: [beta]
+        steps:
+            - setup-ci: git clone https://github.com/screwdriver-cd/toolbox.git ci
+            - wait-docker: DOCKER_TAG=`meta get docker_tag` ./ci/docker-wait.sh
+            - deploy-k8s: K8S_TAG=`meta get docker_tag` ./ci/k8s-deploy.sh
+            # - test: |
```
Comment (jithine) [inline screwdriver.yaml]: we need to have this back as well. let it fail in first run and we can add it progressively

File: plugins/queue/put.js [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+            name: 'Screwdriver scheduler',
+            username: 'sd:scheduler'
+        },
+        causeMessage: causeMessage || 'Automatically started by scheduler'
+    };
+
+    if (eventId) {
+        buildEvent.parentEventId = eventId;
+    }
+    await helper.createBuildEvent(
```
Comment (jithine) [inline plugins/queue/put.js]: we should apply fix from https://github.com/screwdriver-cd/executor-queue/pull/82/files here

File: bin/server (hunk 1 above)
Comment (pritamstyz4ever) [reply bin/server]: Don't need it, will remove

File: screwdriver.yaml (hunk 6 above)
Comment (pritamstyz4ever) [reply screwdriver.yaml]: Yes, adding the tests and test scripts

File: plugins/queue/put.js (hunk 8 above)
Comment (pritamstyz4ever) [reply plugins/queue/put.js]: cool!

File: package.json [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    "eslint-config-screwdriver": "^3.0.1",
+    "mocha": "^7.0.1",
+    "mockery": "^2.1.0",
+    "sinon": "^9.0.0",
+    "util": "^0.12.2"
+  },
+  "scripts": {
+    "start": "node ./bin/server",
+    "pretest": "eslint .",
+    "test": "mocha --recursive --timeout 4000 --retries 1 --exit --allow-uncaught true --color true",
```
Comment (jithine) [inline package.json]: can you use `nyc` as well to collect coverage metric

**Pull request merged**